### PR TITLE
opengl: add support for 16hf formats

### LINKED ISSUE #263: gpu: suboptimal preference of formats - emulated formats can have better performance
In `cmp_fmt()`, non-emulated formats with more caps are always preferred. However, Some GPUs, e.g. my Intel Arc A750, and perhaps other Intel GPUs, have better performance with rgba16f, which is an emulated format, than rgba32f, which is non-emulated. This is confirmed by my test.

Content of my `gpu->formats`:
```
NAME                 TYPE   SIZE COMP CAPS         EMU DEPTH         HOST_BITS     GLSL_TYPE  GLSL_FMT   FOURCC
r8                   UNORM  1    R    SsLRbBV--HWG n   {8  0  0  0 } {8  0  0  0 } float      r8         R8    
rg8                  UNORM  2    RG   SsLRbBV--HWG n   {8  8  0  0 } {8  8  0  0 } vec2       rg8        GR88  
rgba8                UNORM  4    RGBA SsLRbBV--HWG n   {8  8  8  8 } {8  8  8  8 } vec4       rgba8      AB24  
bgra8                UNORM  4    BGRA SsLRbBV--HWG n   {8  8  8  8 } {8  8  8  8 } vec4       rgba8      AR24  
r16                  UNORM  2    R    SsLRbBV--HWG n   {16 0  0  0 } {16 0  0  0 } float      r16        R16   
rg16                 UNORM  4    RG   SsLRbBV--HWG n   {16 16 0  0 } {16 16 0  0 } vec2       rg16       GR32  
rgba16               UNORM  8    RGBA SsLRbBV--HWG n   {16 16 16 16} {16 16 16 16} vec4       rgba16           
r32f                 FLOAT  4    R    SsLRbBV--HWG n   {32 0  0  0 } {32 0  0  0 } float      r32f             
rg32f                FLOAT  8    RG   SsLRbBV--HWG n   {32 32 0  0 } {32 32 0  0 } vec2       rg32f            
rgba32f              FLOAT  16   RGBA SsLRbBV--HWG n   {32 32 32 32} {32 32 32 32} vec4       rgba32f          
r8u                  UINT   1    R    Ss-R-BV--HWG n   {8  0  0  0 } {8  0  0  0 } uint       r8ui             
rg8u                 UINT   2    RG   Ss-R-BV--HWG n   {8  8  0  0 } {8  8  0  0 } uvec2      rg8ui            
rgba8u               UINT   4    RGBA Ss-R-BV--HWG n   {8  8  8  8 } {8  8  8  8 } uvec4      rgba8ui          
r16u                 UINT   2    R    Ss-R-BV--HWG n   {16 0  0  0 } {16 0  0  0 } uint       r16ui            
rg16u                UINT   4    RG   Ss-R-BV--HWG n   {16 16 0  0 } {16 16 0  0 } uvec2      rg16ui           
rgba16u              UINT   8    RGBA Ss-R-BV--HWG n   {16 16 16 16} {16 16 16 16} uvec4      rgba16ui         
r16f                 FLOAT  4    R    SsLRbB---HWG y   {16 0  0  0 } {32 0  0  0 } float      r16f             
rg16f                FLOAT  8    RG   SsLRbB---HWG y   {16 16 0  0 } {32 32 0  0 } vec2       rg16f            
rgba16f              FLOAT  16   RGBA SsLRbB---HWG y   {16 16 16 16} {32 32 32 32} vec4       rgba16f          
rgb8                 UNORM  3    RGB  S-LRbBV--H-G y   {8  8  8  0 } {8  8  8  0 } vec3                  BG24  
rgb16                UNORM  6    RGB  S-LRbBV--H-G y   {16 16 16 0 } {16 16 16 0 } vec3                        
rgb32f               FLOAT  12   RGB  S-LRbBV--H-G y   {32 32 32 0 } {32 32 32 0 } vec3                        
rgb16f               FLOAT  12   RGB  S-LRbB---H-G y   {16 16 16 0 } {32 32 32 0 } vec3                        
rgb8u                UINT   3    RGB  S-----V--H-G y   {8  8  8  0 } {8  8  8  0 } uvec3                       
rgb16u               UINT   6    RGB  S-----V--H-G y   {16 16 16 0 } {16 16 16 0 } uvec3                       
```

It is not strange that even though rgba16f is emulated, it performs better in practice. The GPU can do some internal SIMD with 16f.

### PR DESCRIPTION
**Pull request by ruihe774**
Fixes #263.

### CODE REVIEW

**Comment by haasn** [comment]
Note: We don't currently have a conceivable use case for half float vertex data. It may make sense to reduce the check to just color renderable half float textures (and drop the corresponding V capability), which will also make it a bit less redundant with the existing `formats_float`.

Looks good either way.

**Comment by ruihe774** [comment]
> Note: We don't currently have a conceivable use case for half float vertex data. It may make sense to reduce the check to just color renderable half float textures (and drop the corresponding V capability), which will also make it a bit less redundant with the existing formats_float.

`PL_FMT_CAP_VERTEX` is compared in `cmp_fmt`. Formats without this capacity will not take precedence.

**Comment by haasn** [comment]
https://code.videolan.org/videolan/libplacebo/-/merge_requests/664

**Comment by ruihe774** [comment]
> https://code.videolan.org/videolan/libplacebo/-/merge_requests/664

Rebased.

File: src/opengl/formats.c [hunk 1]
```diff
@@ -124,32 +125,52 @@ const struct gl_format formats_snorm16[] = {
...
-    {GL_RGB16F,         RGB,   FLT, FMT("rgb16f",  16, FLOAT, S|L)},
-    {GL_RGBA16F,        RGBA,  FLT, FMT("rgba16f", 16, FLOAT, S|L)},
+    {GL_R16F,           R,     HALF, FMT("r16hf",   16, FLOAT, S|L)},
+    {GL_RG16F,          RG,    HALF, FMT("rg16hf",  16, FLOAT, S|L)},
+    {GL_RGB16F,         RGB,   HALF, FMT("rgb16hf", 16, FLOAT, S|L)},
+    {GL_RGBA16F,        RGBA,  HALF, FMT("rgba16hf",16, FLOAT, S|L)},
+    {GL_R16F,           R,     FLT,  FMT("r16f",    16, FLOAT, S|L)},
+    {GL_RG16F,          RG,    FLT,  FMT("rg16f",   16, FLOAT, S|L)},
+    {GL_RGB16F,         RGB,   FLT,  FMT("rgb16f",  16, FLOAT, S|L)},
+    {GL_RGBA16F,        RGBA,  FLT,  FMT("rgba16f", 16, FLOAT, S|L)},
```
[the remark this answers is not in the document]
Comment (ruihe774) [reply src/opengl/formats.c:173]: `OES_texture_half_float` and `GL_EXT_color_buffer_half_float` are extensions of OpenGL ES 2.0. We are targeting OpenGL ES 3.0, and `p->gles_ver >= 30` is already checked in that branch.

`GL_EXT_color_buffer_float` is an extension of OpenGL ES 3.0. It makes R16F, RG16F, and RGBA16F renderable. It is absorbed in base OpenGL ES 3.2.

So:

- `gles_ver >= 20`: no 16f formats at all.
- `gles_ver >= 20` + `OES_texture_half_float`: texture-only 16f formats.
- `gles_ver >= 20` + `GL_EXT_color_buffer_half_float`: renderable 16f formats.
- `gles_ver >= 30`: texture-only 16f formats.
- `gles_ver >= 30` + `GL_EXT_color_buffer_float`: renderable 16f formats.
- `gles_ver >= 32`: renderable 16f formats.

Currently we do not enable 16f formats with OpenGL ES 2.0.

[the remark this answers is not in the document]
Comment (haasn) [reply src/opengl/formats.c:173]: Okay, makes sense.

**Comment by haasn** [comment]
It seems that NaNs do not survive an upload/download round-trip on my end. They get consistently written back as 0xFE00.

We could probably modify our test suite to not generate these NaNs in the first place, but given the fact that it works for "native" formats like 32-bit floats, as well as 16-bit floats on Vulkan, I rather suspect that the issue here is that the 16hf code path on my GL driver is actually emulated internally.

Should we mark them as emulated?

```
testing texture roundtrip for format rgba16hf
... 1D
upload time: 2760, download time: 2200
... 2D
upload time: 3000, download time: 2760
... 3D
=== FAILED: memcmp(src, dst, bytes) == 0 at ../src/tests/gpu_tests.h:201
at position 160: 0x0d != 0x00

416 bytes of 'src' at offset 0:
c6 16 21 4c ac b9 32 a8 c5 59 c8 e4 7d f5 12 8c
68 82 3f 33 8e 0f 5a b6 46 38 ae 2a 6c a5 c1 33
bc e3 7f 68 9c b2 10 62 0b d9 46 89 ce 59 15 37
dc 55 6a 6a 65 c4 21 ab fc cf d6 69 74 97 9c 31
7a 1b 9a 17 ce ab 79 d9 85 c0 63 54 1a 79 8b f6
ce f5 60 34 b9 81 df b6 51 b5 1f c6 4d bb f7 c8
d7 92 df a5 3d 59 7e c3 1a e2 17 34 5b a2 2b 2a
98 8c 5e 51 0e 3e 08 5f f4 27 25 41 e2 1d 0a b9
af e9 5e ed 43 dd b0 5e bf c8 92 1b 6b be 45 03
4a a4 55 58 e2 5d b7 d7 84 dc 18 67 fa 23 20 a9
0d 7f 97 50 5d 48 ae 1d 10 41 38 7b ff 7e 7e 49
22 d3 a1 05 31 58 dc b5 35 f5 1d 2f 18 3e d9 25
bd 70 76 1b b8 24 38 c9 66 71 44 65 f0 c2 ae 13
96 50 18 c7 a9 f5 7d df ea 9a 0f 02 d8 e8 28 96
59 9e b2 12 c3 eb db 29 5d 1f 8e 4d e2 3d 60 79
8e 79 40 37 6e be 17 58 59 26 5b 32 0e 84 c9 68
22 7b 7a e5 66 55 0f c3 74 9d 10 56 db 71 d0 6a
ea 10 a1 58 cf b9 b1 28 df 0d 5b ee 91 24 56 b3
9f d0 99 05 25 a9 c9 9a 47 da f1 22 4c c1 8c 36
d2 2e 8f a1 e7 41 ca c6 4e 25 b4 df 49 0b 93 e9
db 2d ee 00 d6 b8 9a 1d 92 8c 40 df 4d cd 15 1f
fb a5 c0 e3 e6 8b aa 35 b0 5f 14 fa 6a a8 e3 45
d6 d2 46 ac 8b e1 ca 1e 6d 0a fd ba d7 12 da d3
b8 9a b7 9e 26 61 d3 d7 c0 e8 d1 2b 91 b5 71 67
88 b7 13 13 99 de 31 06 e8 2e c1 c0 41 9b 93 f9
36 4a 98 5d ac 6c 34 6d 54 06 98 e6 bb 09 4d 44

416 bytes of 'dst' at offset 0:
c6 16 21 4c ac b9 32 a8 c5 59 c8 e4 7d f5 12 8c
68 82 3f 33 8e 0f 5a b6 46 38 ae 2a 6c a5 c1 33
bc e3 7f 68 9c b2 10 62 0b d9 46 89 ce 59 15 37
dc 55 6a 6a 65 c4 21 ab fc cf d6 69 74 97 9c 31
7a 1b 9a 17 ce ab 79 d9 85 c0 63 54 1a 79 8b f6
ce f5 60 34 b9 81 df b6 51 b5 1f c6 4d bb f7 c8
d7 92 df a5 3d 59 7e c3 1a e2 17 34 5b a2 2b 2a
98 8c 5e 51 0e 3e 08 5f f4 27 25 41 e2 1d 0a b9
af e9 5e ed 43 dd b0 5e bf c8 92 1b 6b be 45 03
4a a4 55 58 e2 5d b7 d7 84 dc 18 67 fa 23 20 a9
00 fe 97 50 5d 48 ae 1d 10 41 38 7b 00 fe 7e 49
22 d3 a1 05 31 58 dc b5 35 f5 1d 2f 18 3e d9 25
bd 70 76 1b b8 24 38 c9 66 71 44 65 f0 c2 ae 13
96 50 18 c7 a9 f5 7d df ea 9a 0f 02 d8 e8 28 96
59 9e b2 12 c3 eb db 29 5d 1f 8e 4d e2 3d 60 79
8e 79 40 37 6e be 17 58 59 26 5b 32 0e 84 c9 68
22 7b 7a e5 66 55 0f c3 74 9d 10 56 db 71 d0 6a
ea 10 a1 58 cf b9 b1 28 df 0d 5b ee 91 24 56 b3
9f d0 99 05 25 a9 c9 9a 47 da f1 22 4c c1 8c 36
d2 2e 8f a1 e7 41 ca c6 4e 25 b4 df 49 0b 93 e9
db 2d ee 00 d6 b8 9a 1d 92 8c 40 df 4d cd 15 1f
fb a5 c0 e3 e6 8b aa 35 b0 5f 14 fa 6a a8 e3 45
d6 d2 46 ac 8b e1 ca 1e 6d 0a fd ba d7 12 da d3
b8 9a b7 9e 26 61 d3 d7 c0 e8 d1 2b 91 b5 71 67
88 b7 13 13 99 de 31 06 e8 2e c1 c0 41 9b 93 f9
36 4a 98 5d ac 6c 34 6d 54 06 98 e6 bb 09 4d 44
```

Full log: [testlog.txt](https://github.com/haasn/libplacebo/files/15483687/testlog.txt)

**Comment by ruihe774** [comment]
> It seems that NaNs do not survive an upload/download round-trip on my end. They get consistently written back as 0xFE00.

It is implementation-dependent. e.g. iris fails the test while swrast passes it.

According to OpenGL spec (ver 4.6, sec 2.3.4), passing NaN to GL commands is an UB, despite 16hf or 32f:

> Any representable floating-point value is legal as input to a GL command that requires floating-point data. The result of providing a value that is not a floating-point number to such a command is unspecified, but must not lead to GL interruption or termination. In IEEE arithmetic, for example, providing a negative zero or a denormalized number to a GL command yields predictable results, while providing a NaN or an infinity yields unspecified results.

**Comment by ruihe774** [comment]
BTW it is also an UB in Vulkan.

https://registry.khronos.org/vulkan/specs/1.3/html/chap3.html#fundamentals-general

**Comment by haasn** [comment]
https://code.videolan.org/videolan/libplacebo/-/merge_requests/666

**Comment by haasn** [comment]
Rebased + merged this manually, thanks

**Pull request closed without merging**